engine/gamesys/src/gamesys/resources/res_collision_object.cpp

```diff
@@ -173,51 +177,53 @@ namespace dmGameSystem
+                        resource->m_Shapes2D = (dmPhysics::HCollisionShape2D*)malloc(sizeof(dmPhysics::HCollisionShape2D) * total_shapes_count);
+                        resource->m_ShapeTranslation = (Vectormath::Aos::Vector3*)malloc(sizeof(Vectormath::Aos::Vector3) * total_shapes_count);
+                        resource->m_ShapeRotation = (Vectormath::Aos::Quat*)malloc(sizeof(Vectormath::Aos::Quat) * total_shapes_count);
                         for (uint32_t i = 0; i < shape_count; ++i)
                         {
                             resource->m_Shapes2D[i] = resource->m_TileGridResource->m_GridShapes[i];
+                            resource->m_ShapeTranslation[i] = Vectormath::Aos::Vector3(0.0, 0.0, 0.0);
+                            resource->m_ShapeRotation[i] = Vectormath::Aos::Quat(0.0, 0.0, 0.0, 0.0);
                         }
+                        resource->m_TileGridShapeCount = shape_count;
                         resource->m_ShapeCount = shape_count;
-                        return true;
                     }
-                    // NOTE: Fall-trough "by design" here, ie not "else return false"
-                    // The control flow should be improved. Currently very difficult to follow or understand
-                    // the expected logical paths
                 }
             }
         }
 
-        const dmPhysicsDDF::CollisionShape* embedded_shape = &resource->m_DDF->m_EmbeddedCollisionShape;
-        const dmPhysicsDDF::CollisionShape::Shape* shapes = embedded_shape->m_Shapes.m_Data;
-        if (shapes)
+        if (embedded_shapes)
         {
-            // Create embedded convex shapes
-            uint32_t count = embedded_shape->m_Shapes.m_Count;
-            if (count > COLLISION_OBJECT_MAX_SHAPES)
+            if (physics_context->m_3D)
+            {
+                resource->m_Shapes3D = (dmPhysics::HCollisionShape3D*)malloc(sizeof(dmPhysics::HCollisionShape3D) * embedded_shape_count);
+                resource->m_ShapeTranslation = (Vectormath::Aos::Vector3*)malloc(sizeof(Vectormath::Aos::Vector3) * embedded_shape_count);
+                resource->m_ShapeRotation = (Vectormath::Aos::Quat*)malloc(sizeof(Vectormath::Aos::Quat) * embedded_shape_count);
+            }
+            else if (!resource->m_TileGrid)
             {
-                dmLogWarning("Too many shapes in collision object. Up to %d is supported (%d). Discarding overflowing shapes.", COLLISION_OBJECT_MAX_SHAPES, count);
-                count = COLLISION_OBJECT_MAX_SHAPES;
+                resource->m_Shapes2D = (dmPhysics::HCollisionShape2D*)malloc(sizeof(dmPhysics::HCollisionShape2D) * embedded_shape_count);
+                resource->m_ShapeTranslation = (Vectormath::Aos::Vector3*)malloc(sizeof(Vectormath::Aos::Vector3) * embedded_shape_count);
+                resource->m_ShapeRotation = (Vectormath::Aos::Quat*)malloc(sizeof(Vectormath::Aos::Quat) * embedded_shape_count);
```

Comment:
This code is essentially in 3 places, when 1 should have sufficed.
But let's leave that refactoring until later.

Reply:
Yes, I'm not happy with how it looks.